import1/dbus.go

```diff
@@ -0,0 +1,220 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package import1 provides integration with the systemd-importd API.  See https://www.freedesktop.org/wiki/Software/systemd/importd/
+// Note: Requires systemd v231 or higher
+package import1
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/godbus/dbus"
+)
+
+const (
+	dbusInterface = "org.freedesktop.import1.Manager"
+	dbusPath      = "/org/freedesktop/import1"
+)
+
+// Conn is a connection to systemds dbus endpoint.
+type Conn struct {
+	conn   *dbus.Conn
+	object dbus.BusObject
+}
+
+// Transfer is the status for an import, export or download operations.
+type Transfer struct {
+	Id       uint32  // The numeric transfer ID of the transfer object
+	Local    string  // The local container name of this transfer
+	Remote   string  // The remote source (in case of download: the URL, in case of import/export a string describing the file descriptor passed in)
+	Type     string  // The type of operation
+	Verify   string  // The selected verification setting, and is only defined for download operations
+	Progress float64 // The current progress of the transfer, as a value between 0.0 and 1.0
+}
+
+// New establishes a connection to the system bus and authenticates.
+// Note: systemd-importd will be activated via D-Bus, we don't need to check service status.
+func New() (*Conn, error) {
+	c := new(Conn)
+
+	if err := c.initConnection(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Conn) initConnection() error {
+	var err error
+	c.conn, err = dbus.SystemBusPrivate()
+	if err != nil {
+		return err
+	}
+
+	// Only use EXTERNAL method, and hardcode the uid (not username)
+	// to avoid a username lookup (which requires a dynamically linked
+	// libc)
+	methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(os.Getuid()))}
+
+	err = c.conn.Auth(methods)
+	if err != nil {
+		c.conn.Close()
+		return err
+	}
+
+	err = c.conn.Hello()
+	if err != nil {
+		c.conn.Close()
+		return err
+	}
+
+	c.object = c.conn.Object("org.freedesktop.import1", dbus.ObjectPath(dbusPath))
+
+	return nil
+}
+
+func (c *Conn) getResult(method string, args ...interface{}) (
+	transfer_id uint32, transfer_path dbus.ObjectPath, err error,
+) {
+	result := c.object.Call(fmt.Sprintf("%s.%s", dbusInterface, method), 0, args...)
+	if result.Err != nil {
+		return 0, "", result.Err
+	}
+
+	transfer_id, ok := result.Body[0].(uint32)
+	if !ok {
+		return 0, "", fmt.Errorf("unable to convert dbus response '%v' to uint32", result.Body[0])
+	}
+
+	transfer_path, ok = result.Body[1].(dbus.ObjectPath)
+	if !ok {
+		return 0, "", fmt.Errorf("unable to convert dbus response '%v' to dbus.ObjectPath", result.Body[1])
+	}
+	return
+}
+
+// ImportTar imports a tar into systemd-importd.
+func (c *Conn) ImportTar(
+	fd dbus.UnixFD, local_name string, force, read_only bool,
```

Comment:
Here and below, what do you think of taking an `*os.File` here and doing the fd-conversion internally?

Reply:
Yep, I do agree with this. I will refactor those APIs.